app/controllers/projects/tree_filter_controller.rb

```diff
@@ -0,0 +1,20 @@
+# Controller for viewing a repository's file structure
+class Projects::TreeFilterController < Projects::ApplicationController
+  include ExtractsPath
+  include ActionView::Helpers::SanitizeHelper
+
+  before_action :assign_ref_vars
+
+  def show
+    return render_404 unless @repository.commit(@ref)
+
+    r = Grit::Repo.new(@repo.path_to_repo, { is_bare: true })
+    @tree_filter_list = r.lstree(@ref, recursive: true)

```

Comment:
Can we do this using Rugged? And perhaps over AJAX so we don't slow down the initial request so much, and the JSON body can be cached?